internal/resolution/manifest/maven.go

```diff
@@ -263,20 +264,23 @@ func (m MavenManifestIO) mergeParents(ctx context.Context, result *maven.Project
 		}
 		visited[current.ProjectKey] = true
 		var proj maven.Project
-		if allowLocal && current.RelativePath != "" {
-			currentPath = filepath.Join(filepath.Dir(currentPath), string(current.RelativePath))
-			if filepath.Base(currentPath) != "pom.xml" {
-				// If the base is not pom.xml, this path is a directory but not a file.
-				currentPath = filepath.Join(currentPath, "pom.xml")
-			}
-			f, err := os.Open(currentPath)
+		if parentPath := MavenParentPOMPath(currentPath, string(current.RelativePath)); allowLocal && parentPath != "" {
+			currentPath = parentPath
+			f, err := os.Open(parentPath)
 			if err != nil {
-				return fmt.Errorf("failed to open parent file %s: %w", current.RelativePath, err)
+				return fmt.Errorf("failed to open parent file %s: %w", parentPath, err)
 			}
 			if err := xml.NewDecoder(f).Decode(&proj); err != nil {
 				return fmt.Errorf("failed to unmarshal project: %w", err)
 			}
-		} else {
+			if proj.ProjectKey != current.ProjectKey || proj.Packaging != "pom" {
+				// The identifiers or packaging in parent does not match what we want,
+				// mark proj as empty so parent can be fetched from upstream.
+				proj = maven.Project{}
+			}
+		}
+		// proj being empty indicates that we are not able to find parent pom.xml locally.
+		if reflect.DeepEqual(proj, maven.Project{}) {
```

Comment:
nit: Not sure if I like this `reflect.DeepEqual()` here.
Could maybe instead use a pointer and check `proj == nil` or add a `parentFound bool` and check that?

Though, I'm fine to keep it as-is if you would prefer.

Reply:
actually I considered `parentFound` so just changed the PR to use it